### PR TITLE
Packages: create new react package

### LINKED ISSUE #8159: Packages: create new `react` package
## Feature Description

As per the [package organization doc](https://docs.google.com/document/d/1ImRe9wYKWGCAZiq3jro5KLpSJK_NMKXHC6FWqs3nwl0/edit), this `react` package is intended to be a small wrapper around `use-context-selector` and provides a few useful hooks. It's deliberately not part of the `design-system` package as it is also needed by packages not depending on that.

**References:**

* [Package organization doc](https://docs.google.com/document/d/1ImRe9wYKWGCAZiq3jro5KLpSJK_NMKXHC6FWqs3nwl0/edit#)
* Proof-of-concept implementation: https://github.com/google/web-stories-wp/tree/try/next-level-packages/packages/react/src

**To-do:**

* [ ] Move some of the code from `design-system` and `edit-story` to `/packages/react/src/`  
  This includes but is not limited to:
  * `useBatchingCallback`
  * `useWhyDidYouUpdate` 
  * `createContext` and the like from `use-context-selector`
  * `useContextSelector`
  * `useCombinedRefs`
  * `useReduction`
  * `useUnmount`
  * `usePrevious`
* [ ] Update imports where necessary.  
  Use `@web-stories-wp/...` imports for anything that is already a package, relative imports for everything else.
* [ ] Create `/packages/react/package.json` referencing its dependencies
* [ ] Create `/packages/react/README.md` with short documentation.

**Open question:**

Maybe re-export React itself here and make any package depend on this one. Similar to `@wordpress/element`.

## Additional Context

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

### PR DESCRIPTION
**Pull request by spacedmonkey**
## Summary

Move some of the code from `design-system` and `edit-story` to `/packages/react/src/`  
  This includes but is not limited to:
  * `useBatchingCallback`
  * `useWhyDidYouUpdate` 
  * `createContext` and the like from `use-context-selector`
  * `useContextSelector`
  * `useCombinedRefs`
  * `useReduction`
  * `useUnmount`
  * `usePrevious`
  * `useDebouncedCallback`

### QA

- [x] This is a non-user-facing change and requires no QA

This PR can be tested by following these steps:

1.

### UAT

- [x] UAT should use the same steps as above.

This PR can be tested by following these steps:

1.

## Checklist

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/google/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [x] This PR contains automated tests (unit, integration, and/or e2e) to verify the code works as intended ([docs](https://github.com/google/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Fixes #8159

### CODE REVIEW

File: assets/src/dashboard/app/index.js [hunk 1]
```diff
@@ -31,9 +31,8 @@ import {
   SnackbarProvider,
   Snackbar,
   ModalGlobalStyle,
-  usePrevious,
 } from '@web-stories-wp/design-system';
-
+import { usePrevious } from '@web-stories-wp/react';
```
Comment (swissspidy) [inline assets/src/dashboard/app/index.js:35]: Nit: add empty line after

File: assets/src/dashboard/app/views/editorSettings/publisherLogo/index.js [hunk 2]
```diff
@@ -22,10 +22,9 @@ import { useCallback, useEffect, useRef, useState, useMemo } from 'react';
 import { __ } from '@web-stories-wp/i18n';
 import {
   useGridViewKeys,
-  useFocusOut,
   THEME_CONSTANTS,
 } from '@web-stories-wp/design-system';
-
+import { useFocusOut } from '@web-stories-wp/react';
```
Comment (swissspidy) [inline assets/src/dashboard/app/views/editorSettings/publisherLogo/index.js:27]:
```suggestion
import { useFocusOut } from '@web-stories-wp/react';

```

File: package.json [hunk 3]
```diff
@@ -51,6 +51,7 @@
     "@web-stories-wp/stories-block": "*",
     "@web-stories-wp/stories-carousel": "*",
     "@web-stories-wp/stories-lightbox": "*",
+    "@web-stories-wp/react": "*",
```
Comment (swissspidy) [inline package.json:54]: Can remove the `use-debounce` dependency from this file.

Can also remove `use-reduction` as this doesn't seem to be used anywhere, actually.

File: packages/design-system/src/components/menu/useDropDownMenu.js [hunk 4]
```diff
@@ -13,16 +13,16 @@
...
  * External dependencies
  */
 import { useCallback, useEffect, useMemo, useState } from 'react';
+import { useFocusOut } from '@web-stories-wp/react';
 
 /**
  * Internal dependencies
  */
-import { useFocusOut, isNullOrUndefinedOrEmptyString } from '../../utils';
+import isNullOrUndefinedOrEmptyString from '../../utils/isNullOrUndefinedOrEmptyString';
```
Comment (swissspidy) [inline packages/design-system/src/components/menu/useDropDownMenu.js]:
```suggestion
import { isNullOrUndefinedOrEmptyString } from '../../utils';
```

File: packages/design-system/src/components/search/search.js [hunk 5]
```diff
@@ -21,12 +21,11 @@ import { useCallback, useMemo, useRef } from 'react';
 import PropTypes from 'prop-types';
 import { v4 as uuidv4 } from 'uuid';
 import { __, sprintf } from '@web-stories-wp/i18n';
-
+import { useFocusOut } from '@web-stories-wp/react';
```
Comment (swissspidy) [inline packages/design-system/src/components/search/search.js:24]:
```suggestion
import { useFocusOut } from '@web-stories-wp/react';

```

File: packages/design-system/src/components/textArea/index.js [hunk 6]
```diff
@@ -21,17 +21,14 @@ import PropTypes from 'prop-types';
 import { forwardRef, useMemo, useRef } from 'react';
 import styled, { css } from 'styled-components';
 import { v4 as uuidv4 } from 'uuid';
-
+import { useInputEventHandlers } from '@web-stories-wp/react';
```
Comment (swissspidy) [inline packages/design-system/src/components/textArea/index.js:24]:
```suggestion
import { useInputEventHandlers } from '@web-stories-wp/react';

```

File: packages/react/README.md [hunk 7]
```diff
@@ -0,0 +1,3 @@
+# React
+
+Generic package that re-exports parts of `react` and `react-dom` alongside some other helpful utills.s
```
Comment (swissspidy) [inline packages/react/README.md]:
```suggestion
Generic package that re-exports parts of `react` and `react-dom` alongside some other helpful utils and hooks.
```

File: assets/src/edit-story/components/library/panes/shared/virtualizedPanelGrid/useVirtualizedGridNavigation.js [hunk 8]
```diff
@@ -17,7 +17,8 @@
  * External dependencies
  */
 import { useCallback, useEffect, useMemo, useRef, useState } from 'react';
-import { useFocusOut, useKeyDownEffect } from '@web-stories-wp/design-system';
+import { useKeyDownEffect } from '@web-stories-wp/design-system';
+import { useFocusOut } from '@web-stories-wp/react';
 /**
```
Comment (swissspidy) [inline assets/src/edit-story/components/library/panes/shared/virtualizedPanelGrid/useVirtualizedGridNavigation.js]:
```suggestion
import { useFocusOut } from '@web-stories-wp/react';

/**
```

**Review by swissspidy — APPROVED** [review]
Just a couple of nits

**Comment by swissspidy** [comment]
Some minor merge conflicts now

File: package.json [hunk 9]
```diff
@@ -84,6 +84,7 @@
     "react-virtual": "^2.6.1",
     "sat": "^0.9.0",
     "styled-components": "^5.3.0",
+    "use-debounce": "^7.0.0",
```
Comment (swissspidy) [inline package.json]: Why was this added back?

Comment (spacedmonkey) [reply package.json]: Bad rebase

**Comment by swissspidy** [comment]
> Some minor merge conflicts now

Again 🙈

**Comment by spacedmonkey** [comment]
> Again 🙈

That package.lock is killing me. Merged main now.

**Pull request merged**